cmds/core/cpio/cpio.go

```diff
@@ -75,9 +78,28 @@ func main() {
 			if err != nil {
 				log.Fatalf("error reading records: %v", err)
 			}
-			debug("Creating %s\n", rec)
-			if err := cpio.CreateFile(rec); err != nil {
-				log.Printf("Creating %q failed: %v", rec.Name, err)
+			// The contents of hard linked files are only stored once in the archive. They need special handling.
+			if rec.Mode&0170000 != 0100000 || rec.Info.NLink < 2 {
```

Comment:
I finally figured out what's been bugging me about this. In short, you should not use nlink, ever, 
because on archive creation 
- it could be wrong b/c even if > 2, the file could be one you will not archive
- it could be wrong as it can change between reading the current file and the next file linked to this inode -- files can be added or go away. 

nlink should not be trusted.

so what makes a hard link? The inumber. Further, the very first hard link is the one with the data. IOW you won't have 
zero-length file 
file with data
zero-length file

all sharing an inumber. If you have a set of files with the same inumber, the data-full one always comes first. 

(let's not talk in terms of hard links, it confuses people into thinking one is the "master" copy and the others are somehow shadows or something. Let's just discuss files with same inumber).

So  what you have to do is actually easy.

If the file has data, by definition, you won't have seen it yet. That's how cpio works.

If the file is zero length, you should check the inumber and see if you have already written that inumber. It not, write a zero length file. If so, make a hard link. Done. You won't need the \
// Make sure all hard link files are created. 
loop below. 
so.
file with length of data > 0? write it, and record inumber and path
file with length of data == 0 but an inumber you have not seen yet? create 0-length file
file with length of data == 0 and an inumber you HAVE seen? create a link

that's it.

Reply:
The numbers are a mask (0170000) to get rid of the file type bits. For what's left, value 0100000 is a regular file (not a directory). I got them from pkg/cpio/fs_unix.go where they are nicely defined as constants, but unexported. Would it be better to redefine them here as constants?

Reply:
I was going by the the cpio 'New ASCII' format documentation in cpio.formats in this directory, which states:

     In this format, hardlinked files are handled by setting the filesize to
     zero for each entry except the last one that appears in the archive.

I thought it seemed backwards to do it that way... And it appears to be not how cpio actually does it. The entry with data is actually the first one in the archive, just like you said it would be.

You're right that relying on nlink is a problem. I hadn't considered the race scenario.

I've pushed an updated version that does things the simple way and no longer relies on nlink.